# Student Performance Indicator

### Life cycle of Machine Learning Project

- Understanding the problem statement
- Data collection
- Data checks to perform
- Exploratory data analysis
- Data Pre-processing
- Model Training
- choose best model 

## 1. Problem Statement

This project understands how the student's performance(test scores) is affected by other variables such as Gender, Ethnicity, Parental level of Educaiton, Lunch and Test preparation course.

## 2. Data Collection
Data Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?resource=download

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

In [5]:
df = pd.read_csv('data/StudentsPerformance.csv')

In [6]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [7]:
df.shape

(1000, 8)

## Data Checks to perform

- Check missing values
- Check Duplicate
- Check data type
- Check the number of unique values of each column
- Check statistics of data set
- Check various categories present in different categorical column

In [8]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

In [9]:
df.duplicated().sum()

0

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [11]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

In [13]:
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

## Insights

- From above decription of numerical data, all means are very close to each other - b/w 66 and 69.169
- All standard deviations are also cloase - b/w 14.6 amd 15.19
- While there is a minimum score 0 for math , reading is 17 and writing 10

In [15]:
print("Categories in gender variable : ",end=" ")
print(df['gender'].unique())

print("Categories in race/ethnicity variable : ",end=" ")
print(df['race/ethnicity'].unique())

print("Categories in parental level of education variable : ",end=" ")
print(df['parental level of education'].unique())

print("Categories in lunch variable : ",end=" ")
print(df['lunch'].unique())

print("Categories in test preparation course variable : ",end=" ")
print(df['test preparation course'].unique())

Categories in gender variable :  ['female' 'male']
Categories in race/ethnicity variable :  ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in parental level of education variable :  ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in lunch variable :  ['standard' 'free/reduced']
Categories in test preparation course variable :  ['none' 'completed']


In [17]:
numerical_features = [feature for feature in df.columns if df[feature].dtype != 'O' ]
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O' ]

print('We have {} numerical features : {}'.format(len(numerical_features),numerical_features))
print('We have {} categorical features : {}'.format(len(categorical_features),categorical_features))

We have 3 numerical features : ['math score', 'reading score', 'writing score']
We have 5 categorical features : ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']


## Adding columns for 'Total Score' and 'Average'

In [19]:
df['total score'] = df['math score']+df['writing score']+df['reading score']
df['average'] = df['total score']/3
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total score    average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333

In [ ]:
reading_full = df[]